### ISSUE TITLE
What could be causing horrible performance when running in a VM?

### ISSUE THREAD

**Issue by mnmr** [microsoft/FASTER#379]
We're trying to figure out a problem that only occurs when we run our code in an Azure VM. As soon as we spawn the tasks to create and initialize the Faster stores, CPU usage goes up to 100% and the process becomes unresponsive. We have about 1000 stores but they are all really tiny (less than 1000 messages per store).

I've captured a few performance snapshots to hopefully provide some pointers as to what the problem could be:

![image (3)](https://user-images.githubusercontent.com/1616137/100405283-94b94f00-3063-11eb-83de-226a8162eac4.png)

![image (2)](https://user-images.githubusercontent.com/1616137/100405267-8a975080-3063-11eb-8618-0972aa262b95.png)

This is the code we use to create FasterLog:

            var fileName = Path.Combine(options.Path, "store");
            device = Devices.CreateLogDevice(fileName, false, false, -1L, false);

            // 2^PSB per page, 2^(MSB-PSB) pages in mem, 2^SSB per disk segment/file
            // 32kb, 4 pages, 1gb disk
            faster = new FasterLog(new FasterLogSettings
            {
                LogDevice = device,
                ReadOnlyMode = readOnly,
                MutableFraction = 0,
                PageSizeBits = 15,
                MemorySizeBits = 17,
                SegmentSizeBits = 30
            });

We've excluded the folder where the stores are kept from the Windows anti-virus scanner, but that had no discernible effect.

Any ideas as to what the problem could be?

**Comment by badrishc** [comment]
Does it spin forever? If not, how long does it spin?

 I assume this happens only when starting (recovering) and not during runtime?

**Comment by mnmr** [comment]
> Does it spin forever? If not, how long does it spin?

It's not forever, as the first few stores take 5-10 seconds to open. Times then rapidly increase as the system attempts to open more stores and the process does not recover after that (I've left it for more than 5 minutes and it had not made significant progress at that point).
 
> I assume this happens only when starting (recovering) and not during runtime?

Yes, if I remove all existing stores then startup is instant and the system hums along nicely. The problem occurs when the service is restarted, e.g. when we deploy an update.

**Comment by badrishc** [comment]
The latest in master no longer spins during recovery. Instead it waits on a semaphore. Wondering if you could try that out.

**Comment by mnmr** [comment]
> The latest in master no longer spins during recovery. Instead it waits on a semaphore. Wondering if you could try that out.

This fixes the CPU usage, but unfortunately does nothing for the startup times:

![image](https://user-images.githubusercontent.com/1616137/100490484-005ef300-311c-11eb-916b-d17c0a129cf2.png)

It's odd that even stores with 0 messages don't initialize quickly. It also appears as if stores finish recovery in batches, rather than independently. Do all the stores share the same semaphore? 

After about 4 more minutes I got a major log dump with what looked like the rest of the stores finishing (with init times from 4m up to 6m). All of this takes just seconds on my local development PC, and I suspect console logging is what slows the process down the most there.

**Comment by badrishc** [comment]
Nothing comes to mind based on the description. Is the 1000 stores just some kind of stress test? There might be some way to multitask the contents into a fewer number of logs in practice? The random IO might be stressful if there is a lot of activity across the stores.

**Comment by mnmr** [comment]
> Nothing comes to mind based on the description. Is the 1000 stores just some kind of stress test? There might be some way to multitask the contents into a fewer number of logs in practice? The random IO might be stressful if there is a lot of activity across the stores.

We don't do anything fancy at all. Each store is owner by an actor (so is only ever accessed from one thread). During startup the entire store is read sequentially (and processed to recreate the current actor state), and after that it is only ever used as an append log. 

We use a single store per actor as that makes it easy to clean up the data when it is no longer needed. The largest store is about 1mb and most of the stores are <100kb, for a total of some 200mb across 1k stores. Even so, it's the mere creation of the FasterLog device that consumes all of the time (I abort the data scan after 30s as a fail-safe, but am not seeing the corresponding logs to indicate that that is happening). I've also tried toggling the "recoverOnOpen" option passed when creating the log device, but that had no discernible effect.

**Comment by badrishc** [comment]
Any chance you could create a simple stand alone repro? We can run it in a VM and see.

**Comment by badrishc** [comment]
Each store has its own separate commit folder, correct?

**Comment by TedHartMS** [comment]
What's the VM configuration?

**Comment by badrishc** [comment]
Another thought: can you try to eliminate `EnableProcessPrivileges` by just returning false in this method: https://github.com/microsoft/FASTER/blob/3853e88fc48758e557e3139cd4b12593b136d397/cs/src/core/Utilities/Native32.cs#L273

**Comment by mnmr** [comment]
> Any chance you could create a simple stand alone repro? We can run it in a VM and see.

Good idea, I'll create one. [update: maybe not, after having tried the latest suggestion; let me know if you still need one]

> Each store has its own separate commit folder, correct?

Yes, each store lives in a folder of its own.

> What's the VM configuration?

![image](https://user-images.githubusercontent.com/1616137/100525507-b9c3d400-31c1-11eb-9e6a-867e1f9667ab.png)
![image](https://user-images.githubusercontent.com/1616137/100525530-ee379000-31c1-11eb-86a4-83cc9447883d.png)

> Another thought: can you try to eliminate EnableProcessPrivileges by just returning false in this method

THIS makes a huuge difference! All stores now completely initialize within a minute, and init times are more or less correlated with the number of messages in the store.

**Comment by badrishc** [comment]
That's awesome (no need for a repro now). There seems to be some scalability inefficiency with the system calls in `EnableProcessPrivileges` and the Azure VMs. We will try to get a better understanding and fix. Worst case, this can be a configurable feature.

**Comment by badrishc** [comment]
@mnmr - trying to diagnose this. Can you try the following:
* Undo the change you made to `EnableProcessProvileges`
* Return false at the beginning of `EnableVolumePrivileges` here : https://github.com/microsoft/FASTER/blob/3853e88fc48758e557e3139cd4b12593b136d397/cs/src/core/Utilities/Native32.cs#L309

And let us know if that also resolves the issue.

**Comment by badrishc** [comment]
The linked PR should fix the issue.

* As long as `preallocateFile` is set to false (when creating the device), you will not need to do anything else special in your code.
* If you do set `preallocateFile` to true, then FASTER will try to set privileges by default. To disable this behavior, you can use this static configuration setting:
   ```cs
   LocalStorageDevice.UsePrivileges = false;
   ```

Note: The reason for trying to set privileges is disk performance. On a local NVMe device, there is around 20% speed to be gained by using privileges to set the valid file size, when `preallocateFile` is enabled.
